src/overlays/actors/ovl_En_Si/z_en_si.c

```diff
@@ -77,24 +90,81 @@ static DamageTable D_8098CDB8 = {
     /* Powder Keg     */ DMG_ENTRY(1, 0x0),
 };
 
-#endif
-
-extern ColliderSphereInit D_8098CD80;
-extern CollisionCheckInfoInit2 D_8098CDAC;
-extern DamageTable D_8098CDB8;
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/func_8098CA20.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/func_8098CAD0.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/func_8098CB70.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/func_8098CBDC.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/EnSi_Init.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/EnSi_Destroy.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/EnSi_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Si/EnSi_Draw.s")
+void func_8098CA20(EnSi* this, PlayState* play) {
+    this->collider_sphere.dim.worldSphere.center.x = (s16)(s32)this->actor.world.pos.x;
+    this->collider_sphere.dim.worldSphere.center.y = (s16)(s32)this->actor.world.pos.y;
+    this->collider_sphere.dim.worldSphere.center.z = (s16)(s32)this->actor.world.pos.z;
+    this->collider_sphere.dim.worldSphere.radius =
+        (s16)(s32)((f32)this->collider_sphere.dim.modelSphere.radius * this->collider_sphere.dim.scale);
+    if ((s32)this->actor.colChkInfo.health > 0) {
+        CollisionCheck_SetAC(play, &play->colChkCtx, &this->collider_sphere.base);
+    }
+    CollisionCheck_SetOC(play, &play->colChkCtx, &this->collider_sphere.base);
+}
+
+void func_8098CAD0(EnSi* this, PlayState* play) {
+    s32 temp_a2;
+
+    temp_a2 = (s32)(this->actor.params & 0xFC) >> 2;
+    if ((temp_a2 < 0x20) && (temp_a2 >= 0)) {
+        Flags_SetTreasure(play, temp_a2);
+    }
+    Item_Give(play, ITEM_SKULL_TOKEN);
+    if (Inventory_GetSkullTokenCount(play->sceneId) >= 0x1E) {
+        Message_StartTextbox(play, 0xFCU, NULL);
+        Audio_PlayFanfare(NA_SE_PL_DUMMY_290);
+        return;
+    }
+    Message_StartTextbox(play, 0x52U, NULL);
+    Audio_PlayFanfare(NA_BGM_GET_SMALL_ITEM);
+}
+
+void func_8098CB70(EnSi* this, PlayState* play) {
+    if ((this->actor.flags & 0x2000) == 0x2000) {
+        this->actionFunc = func_8098CBDC;
+        goto block_4;
```

Comment:
I'll have to take a close look at this when I get a little more time, but this goto can probably be removed with adjusting the control flow a bit.

Reply:
Not sure how without affecting the resulting ROM. For example a little reordering should not affect the functionality although it affects the resulting ASM:
```c
void func_8098CB70(EnSi* this, PlayState* play) {
    if (this->collider_sphere.base.ocFlags2 & OC2_HIT_PLAYER) {
        func_8098CAD0(this, play);
        Actor_Kill(&this->actor);
        return;
    }
    if ((this->actor.flags & 0x2000) == 0x2000) {
        this->actionFunc = func_8098CBDC;
    }
    this->actor.shape.rot.y += 0x38E;
}
```

Reply:
Maybe an `else` would do the trick?
```c
    if ((this->actor.flags & 0x2000) == 0x2000) {
        this->actionFunc = func_8098CBDC;
    } else if (this->collider_sphere.base.ocFlags2 & OC2_HIT_PLAYER) {
        func_8098CAD0(this, play);
        Actor_Kill(&this->actor);
        return;
    }
    this->actor.shape.rot.y += 0x38E;
```

Reply:
Haha nice one! I updated it